File: src/issue_orchestrator/view_models/tech_lead_activity.py
"""Projection of the local tech-lead run history onto the dashboard (#6858).

ADR-0033's second surface. :mod:`.tech_lead_run_actions` projects what the
operator can DO right now; this projects what the tech lead has already done —
the "tech-lead activity" view the ADR names, and the thing that makes a run a
visible object rather than a session on a bookkeeping issue nobody reads.

Presentation only. Nothing here decides anything: the phases come from the
recorded run, and every string this emits is rendered verbatim so the surface
is legible without a colour key (the same rule the run-action affordances
follow). The subject is shown as a REFERENCE — "#123", a link out — never as
the run's own identity, because on a client's board that issue belongs to them.
"""

from __future__ import annotations

from typing import TYPE_CHECKING, Union

from pydantic import BaseModel, ConfigDict, Field

from ..domain.tech_lead_run_artifacts import TechLeadRunArtifactKind
from ..domain.tech_lead_run_record import TechLeadRunPhase, TechLeadRunSubjectKind
from ..domain.tech_lead_session import TechLeadSessionFlavor
from .lifecycle_semantics import (
    OpenReviewArtifactCommand,
    OpenSessionRecordingCommand,
)

if TYPE_CHECKING:
    from ..domain.tech_lead_run_artifacts import TechLeadRunArtifacts
    from ..domain.tech_lead_run_record import TechLeadRunRecord
    from ..ports.tech_lead_run_record_store import TechLeadRunHistoryReader


# How many runs the panel shows. Small on purpose: this is a "what has the tech
# lead been doing?" glance, not an audit log — the full history stays queryable
# in the local store.
ACTIVITY_LIMIT = 20

# Human names for the three run flavors. Here rather than in the browser so the
# vocabulary has one owner: a flavor renamed in the domain cannot leave the UI
# quietly showing the old word.
_FLAVOR_LABELS: dict[TechLeadSessionFlavor, str] = {
    TechLeadSessionFlavor.HEALTH_REVIEW: "Health review",
    TechLeadSessionFlavor.BATCH_REVIEW: "Batch review",
    TechLeadSessionFlavor.FAILURE_INVESTIGATION: "Failure investigation",
}

# Colour-independent phase words, per the project's rule that status must never
# be signalled by tint alone.
_PHASE_LABELS: dict[TechLeadRunPhase, str] = {
    TechLeadRunPhase.RUNNING: "Running",
    TechLeadRunPhase.COMPLETED: "Completed",
    TechLeadRunPhase.NEEDS_HUMAN: "Needs human",
    TechLeadRunPhase.FAILED: "Failed",
    TechLeadRunPhase.WITHDRAWN: "Withdrawn",
}

# The tone each phase renders with. A CLASS name, not a colour: the stylesheet
# owns the palette, and both themes get the same semantic buckets.
_PHASE_TONES: dict[TechLeadRunPhase, str] = {
    TechLeadRunPhase.RUNNING: "active",
    TechLeadRunPhase.COMPLETED: "good",
    TechLeadRunPhase.NEEDS_HUMAN: "warn",
    TechLeadRunPhase.FAILED: "bad",
    TechLeadRunPhase.WITHDRAWN: "muted",
}


# What a whole-repository run's subject is CALLED. A global run has no subject
# issue (#6858 F5), so without a word for the board and the PR manifest the panel
# would either show nothing or show the coordination anchor as if it were the
# subject — which is the confusion the record was fixed to stop.
_SUBJECT_LABELS: dict[TechLeadRunSubjectKind, str] = {
    TechLeadRunSubjectKind.BOARD: "Whole board",
    TechLeadRunSubjectKind.PR_MANIFEST: "PR manifest",
}

# Operator-facing name for each drill-down. Server-owned for the same reason
# phase words are: the browser must not invent a second vocabulary.
_ARTIFACT_LABELS: dict[TechLeadRunArtifactKind, str] = {
    TechLeadRunArtifactKind.SESSION_REPLAY: "Session replay",
    TechLeadRunArtifactKind.REPORT: "Report",
    TechLeadRunArtifactKind.DECISION: "Decision",
}

# Why a run offers no drill-down. Two different facts, so two sentences: one is
# "not yet", the other is "never", and an operator needs to tell them apart.
ARTIFACTS_PENDING_NOTE = "Artifacts are preserved when the run ends."
ARTIFACTS_ABSENT_NOTE = "No artifacts were preserved for this run."

# The commands a recorded run publishes. Deliberately the EXISTING lifecycle
# inspection commands rather than a tech-lead-shaped copy of them: the dashboard
# already has one dispatcher for "open this run's recording / artifact", and a
# parallel vocabulary would mean a second dispatcher and a second place for the
# artifact-read contract to drift (#6858 round 1 A1/F4).
TechLeadRunArtifactCommand = Union[
    OpenSessionRecordingCommand, OpenReviewArtifactCommand
]


class TechLeadRunActivityEntry(BaseModel):
    """One recorded run, as the activity panel renders it."""

    model_config = ConfigDict(extra="forbid", frozen=True, populate_by_name=True)

    run_key: str = Field(serialization_alias="runKey")
    # "Health review" / "Batch review" / "Failure investigation".
    flavor_label: str = Field(serialization_alias="flavorLabel")
    phase: str
    phase_label: str = Field(serialization_alias="phaseLabel")
    # Semantic bucket for styling. Never the only signal — ``phase_label`` is
    # always rendered too.
    tone: str
    started_at: str = Field(serialization_alias="startedAt")
    # "" while the run is still going.
    ended_at: str = Field(serialization_alias="endedAt")
    # What the run is ABOUT: "issue" / "board" / "pr_manifest".
    subject_kind: str = Field(serialization_alias="subjectKind")
    # The rendered subject — "#42 Flaky merge queue", "Whole board", "PR
    # manifest". One sentence per subject kind, owned here so the browser cannot
    # start calling a board review an investigation of its anchor.
    subject_label: str = Field(serialization_alias="subjectLabel")
    # 0 for every whole-repository run: its subject is not an issue (#6858 F5).
    subject_issue_number: int = Field(serialization_alias="subjectIssueNumber")
    subject_title: str = Field(serialization_alias="subjectTitle")
    # The bookkeeping issue the run was COORDINATED through (0 when none).
    # Published separately from the subject so an operator can still reach the
    # anchor without the panel claiming the run was about it.
    anchor_issue_number: int = Field(serialization_alias="anchorIssueNumber")
    detail: str
    findings: int
    proposals: int
    # Drill-down identity: what the session-replay surface needs to find this
    # run's transcript. Published rather than reconstructed in the browser so
    # the UI never has to know how run artifacts are addressed.
    run_id: str = Field(serialization_alias="runId")
    session_name: str = Field(serialization_alias="sessionName")
    # The inspections available for this run, in one fixed order. Empty when the
    # run is still going or preserved nothing.
    artifacts: tuple[TechLeadRunArtifactCommand, ...]
    # Why there are none ("" when there are some).
    artifacts_note: str = Field(serialization_alias="artifactsNote")


class TechLeadActivityView(BaseModel):
    """The dashboard's tech-lead activity panel."""

    model_config = ConfigDict(extra="forbid", frozen=True, populate_by_name=True)

    entries: tuple[TechLeadRunActivityEntry, ...]
    # The sentence shown when there are none. Published rather than hardcoded in
    # the browser: "nothing has run yet" and "this engine has no history" read
    # the same to the UI but not to an operator.
    empty_message: str = Field(serialization_alias="emptyMessage")

    @classmethod
    def empty(cls) -> "TechLeadActivityView":
        return cls(entries=(), empty_message=EMPTY_MESSAGE)


EMPTY_MESSAGE = "No tech-lead runs recorded yet."


def read_tech_lead_activity(
    history: "TechLeadRunHistoryReader", *, limit: int = ACTIVITY_LIMIT
) -> TechLeadActivityView:
    """Project the local run history onto the activity panel."""
    entries = tuple(_entry(record) for record in history.recent(limit=limit))
    return TechLeadActivityView(entries=entries, empty_message=EMPTY_MESSAGE)


def _entry(record: "TechLeadRunRecord") -> TechLeadRunActivityEntry:
    commands = _artifact_commands(record)
    return TechLeadRunActivityEntry(
        run_key=record.run_key,
        flavor_label=_FLAVOR_LABELS[record.flavor],
        phase=record.phase.value,
        phase_label=_PHASE_LABELS[record.phase],
        tone=_PHASE_TONES[record.phase],
        started_at=record.started_at.isoformat(),
        ended_at=record.ended_at.isoformat() if record.ended_at else "",
        subject_kind=record.subject_kind.value,
        subject_label=_subject_label(record),
        subject_issue_number=record.subject_issue_number,
        subject_title=record.subject_title,
        anchor_issue_number=record.anchor_issue_number,
        detail=record.detail,
        findings=record.findings,
        proposals=record.proposals,
        run_id=record.run_id,
        session_name=record.session_name,
        artifacts=commands,
        artifacts_note=_artifacts_note(record, commands),
    )


def _subject_label(record: "TechLeadRunRecord") -> str:
    """What this run is about, in words.

    A focused investigation names its issue as a REFERENCE ("#42 …"); a global
    run names the board or the manifest, because that is its subject and its
    anchor is only how it was coordinated.
    """
    if record.subject_kind is TechLeadRunSubjectKind.ISSUE:
        title = f" {record.subject_title}" if record.subject_title else ""
        return f"#{record.subject_issue_number}{title}"
    return _SUBJECT_LABELS[record.subject_kind]


def _artifact_commands(
    record: "TechLeadRunRecord",
) -> tuple[TechLeadRunArtifactCommand, ...]:
    """The inspections this run's PRESERVED artifacts support.

    Driven entirely by what the archive actually filed: a run whose decision was
    never written offers no Decision button, so the panel cannot present an
    action that 404s.
    """
    artifacts: "TechLeadRunArtifacts | None" = record.artifacts
    if artifacts is None:
        return ()
    issue_number = record.subject_issue_number or record.anchor_issue_number
    run_dir = str(artifacts.location)
    return tuple(
        _command_for(kind, issue_number=issue_number, run_dir=run_dir)
        for kind in artifacts.kinds
    )


def _command_for(
    kind: TechLeadRunArtifactKind, *, issue_number: int, run_dir: str
) -> TechLeadRunArtifactCommand:
    """The typed inspection command for one preserved artifact kind.

    ``artifact_path`` is run-RELATIVE and comes from the domain's kind→member
    table, so the run-scoped artifact reader resolves and contains it against the
    preserved directory rather than trusting a path the browser assembled.
    """
    if kind is TechLeadRunArtifactKind.REPORT:
        return OpenReviewArtifactCommand(
            label=_ARTIFACT_LABELS[kind],
            issue_number=issue_number,
            run_dir=run_dir,
            artifact_path=str(kind.member),
            artifact_type="tech_lead_report",
            render_mode="markdown",
        )
    if kind is TechLeadRunArtifactKind.DECISION:
        return OpenReviewArtifactCommand(
            label=_ARTIFACT_LABELS[kind],
            issue_number=issue_number,
            run_dir=run_dir,
            artifact_path=str(kind.member),
            artifact_type="tech_lead_decision",
            render_mode="json",
        )
    return OpenSessionRecordingCommand(
        label=_ARTIFACT_LABELS[kind],
        issue_number=issue_number,
        run_dir=run_dir,
    )


def _artifacts_note(
    record: "TechLeadRunRecord",
    commands: tuple[TechLeadRunArtifactCommand, ...],
) -> str:
    if commands:
        return ""
    return (
        ARTIFACTS_ABSENT_NOTE if record.phase.is_terminal else ARTIFACTS_PENDING_NOTE
    )


__all__ = [
    "ACTIVITY_LIMIT",
    "ARTIFACTS_ABSENT_NOTE",
    "ARTIFACTS_PENDING_NOTE",
    "EMPTY_MESSAGE",
    "TechLeadActivityView",
    "TechLeadRunActivityEntry",
    "TechLeadRunArtifactCommand",
    "read_tech_lead_activity",
]

File: src/issue_orchestrator/static/js/dashboard/tech_lead_activity.js
// Tech-lead activity panel — ADR-0033's local visibility surface (#6858).
//
// ``tech_lead_runs.js`` owns what the operator can DO; this owns what the tech
// lead has already DONE. It renders the typed ``techLeadActivity`` payload the
// dashboard view model publishes: one row per recorded run, with its scope, its
// phase, what it produced, and the subject it REFERENCED (never owned).
//
// Nothing is decided here. Phase words, tones, flavor names and the empty-state
// sentence all arrive from the server projection, so the browser cannot invent
// a second vocabulary for run state — the same rule the run-action affordances
// follow.
//
// Accessibility:
//
// * The panel is a native <details> disclosure with a real <summary>, so it is
//   keyboard reachable, has an accessible name, and reports expanded state
//   without any ARIA of our own.
// * Runs are a real <ul>/<li> list, so a screen reader announces how many there
//   are and where it is in them.
// * Phase is ALWAYS rendered as text (``phaseLabel``) next to its tone class —
//   colour is never the only signal.
// * Drill-downs are native <button> elements with real text labels, rendered by
//   the shared lifecycle-Command owner, so they are keyboard reachable and carry
//   the same visible focus ring as every other dashboard action.
// * Refreshes RECONCILE keyed rows: unchanged rows are never rewritten, so an
//   operator tabbing through the panel keeps their place. When a row's markup
//   does change, focus is returned to the same action in the rebuilt row. The
//   <details>/<summary> nodes are never replaced at all.

function techLeadActivityState() {
    return (window.dashboardData && window.dashboardData.techLeadActivity) || null;
}

// Pure: the runs to render, newest first (the server already ordered them).
function techLeadActivityEntries(activity) {
    return (activity && Array.isArray(activity.entries)) ? activity.entries : [];
}

// Pure: the run's subject, as the server named it — "#123 Some title" for a
// focused investigation, "Whole board" / "PR manifest" for a global run. The
// browser never derives this: a whole-board review's subject is NOT the anchor
// it was coordinated through, and deciding that here is how the two halves of
// ADR-0033 got confused in the first place.
function techLeadActivitySubjectText(entry) {
    return entry.subjectLabel ? String(entry.subjectLabel) : '';
}

// Pure: "via #900", or '' when the anchor IS the subject (or there is none).
// Shown so an operator can still reach a global run's bookkeeping issue without
// the panel claiming the run was about it.
function techLeadActivityAnchorText(entry) {
    const anchor = Number(entry.anchorIssueNumber) || 0;
    const subject = Number(entry.subjectIssueNumber) || 0;
    if (!anchor || anchor === subject) {
        return '';
    }
    return `via #${anchor}`;
}

// Pure: the run's drill-down buttons, or the server's sentence explaining why
// there are none. Every button is a typed lifecycle Command the server built
// from the PRESERVED artifact location, rendered and dispatched by the shared
// command owner — so this panel never assembles an endpoint, a path, or a
// fallback guess from runId/sessionName.
function techLeadActivityActionsHtml(entry) {
    const commands = Array.isArray(entry.artifacts) ? entry.artifacts : [];
    if (!commands.length) {
        const note = entry.artifactsNote ? String(entry.artifactsNote) : '';
        return note
            ? `<span class="tla-artifacts-note">${escapeHtml(note)}</span>`
            : '';
    }
    // Each control sits in a slot carrying the run + action identity. That is
    // what lets a refresh put keyboard focus back on the SAME action after a row
    // is re-rendered, instead of dropping it to the document (#6858 F11).
    const key = techLeadActivityRowKey(entry);
    const buttons = commands
        .map((command, index) => (
            `<span class="tla-action-slot" data-tla-row="${escapeAttr(key)}"`
            + ` data-tla-action="${index}">`
            + _renderLifecycleCommandButton(command, null, 'issue-action-btn tla-action')
            + '</span>'
        ))
        .join('');
    return `<span class="tla-actions">${buttons}</span>`;
}

// Pure: the identity of one recorded run's row. The session run pair is the
// record's own primary key, so a row keeps its identity across refreshes even as
// its phase, counts and drill-downs change.
function techLeadActivityRowKey(entry) {
    return `${entry.runId || ''}::${entry.sessionName || ''}`;
}

// Pure: "2 findings · 1 proposal", or '' when the run produced neither.
// Singular/plural is decided here rather than server-side because it is pure
// presentation of two numbers the projection already publishes.
function techLeadActivityProducedText(entry) {
    const parts = [];
    const findings = Number(entry.findings) || 0;
    const proposals = Number(entry.proposals) || 0;
    if (findings) {
        parts.push(`${findings} finding${findings === 1 ? '' : 's'}`);
    }
    if (proposals) {
        parts.push(`${proposals} proposal${proposals === 1 ? '' : 's'}`);
    }
    return parts.join(' · ');
}

// Pure: one <li> for a recorded run. Every interpolated value is escaped: the
// subject title and the decision summary are agent- and GitHub-authored text.
function renderTechLeadActivityRow(entry) {
    const cells = [
        `<span class="tla-flavor">${escapeHtml(entry.flavorLabel || '')}</span>`,
        `<span class="tla-phase tla-phase--${escapeAttr(entry.tone || 'muted')}">`
        + `${escapeHtml(entry.phaseLabel || '')}</span>`,
    ];
    const subject = techLeadActivitySubjectText(entry);
    if (subject) {
        cells.push(`<span class="tla-subject">${escapeHtml(subject)}</span>`);
    }
    const anchor = techLeadActivityAnchorText(entry);
    if (anchor) {
        cells.push(`<span class="tla-anchor">${escapeHtml(anchor)}</span>`);
    }
    const started = formatTimestamp(entry.startedAt, '');
    if (started) {
        cells.push(`<span class="tla-started">${escapeHtml(started)}</span>`);
    }
    const produced = techLeadActivityProducedText(entry);
    if (produced) {
        cells.push(`<span class="tla-produced">${escapeHtml(produced)}</span>`);
    }
    if (entry.detail) {
        cells.push(
            `<span class="tla-detail" title="${escapeAttr(entry.detail)}">`
            + `${escapeHtml(entry.detail)}</span>`,
        );
    }
    const actions = techLeadActivityActionsHtml(entry);
    if (actions) {
        cells.push(actions);
    }
    const key = escapeAttr(techLeadActivityRowKey(entry));
    return `<li class="tla-row" data-tla-row="${key}">${cells.join('')}</li>`;
}

// Pure: the keyed rows to render, newest first. The empty state is a row too, so
// the reconciler below has one shape to compare against.
function techLeadActivityRowModels(activity) {
    const entries = techLeadActivityEntries(activity);
    if (!entries.length) {
        const message = (activity && activity.emptyMessage) || '';
        return [{
            key: '',
            html: `<li class="tla-row tla-row--empty" data-tla-row="">`
                + `${escapeHtml(message)}</li>`,
        }];
    }
    return entries.map(entry => ({
        key: techLeadActivityRowKey(entry),
        html: renderTechLeadActivityRow(entry),
    }));
}

// Pure: the whole list body — rows, or the server's empty-state sentence.
function techLeadActivityRowsHtml(activity) {
    return techLeadActivityRowModels(activity).map(row => row.html).join('');
}

// Pure: how to get from the rows currently in the DOM to the ones the payload
// asks for. ``rebuild`` means the row SET changed (a run started, one aged out
// of the window) and the list is written wholesale; otherwise only the rows whose
// markup actually differs are replaced, so unchanged rows — and any focus inside
// them — are never touched (#6858 F11).
function techLeadActivityRowPlan(activity, existingKeys) {
    const rows = techLeadActivityRowModels(activity);
    const current = Array.isArray(existingKeys) ? existingKeys : [];
    const aligned = current.length === rows.length
        && rows.every((row, index) => current[index] === row.key);
    if (!aligned) {
        return { rebuild: true, html: rows.map(row => row.html).join(''), replace: [] };
    }
    return { rebuild: false, html: '', replace: rows };
}

// Pure: which artifact action holds keyboard focus, as {row, action} — or null
// when focus is elsewhere. Walks up from the focused control to its slot, so it
// works whether the browser focused the button or something inside it.
function techLeadActivityFocusedAction(active) {
    let node = active;
    for (let depth = 0; node && depth < 4; depth += 1) {
        const data = node.dataset;
        if (data && data.tlaAction !== undefined && data.tlaRow !== undefined) {
            return { row: data.tlaRow, action: data.tlaAction };
        }
        node = node.parentElement;
    }
    return null;
}

// Pure: the count shown beside the panel's name.
function techLeadActivityCountText(activity) {
    const count = techLeadActivityEntries(activity).length;
    return count ? String(count) : '';
}

// Surgically update the panel. The <details>/<summary> nodes are never replaced,
// unchanged rows are never rewritten, and a focused drill-down is returned to
// the same action when its row IS rewritten — so an operator tabbing through an
// expanded panel keeps both the panel and their place in it across every live
// SSE-driven refresh.
function updateTechLeadActivityPanel(activity) {
    const panel = document.getElementById('techLeadActivityPanel');
    // No payload yet is NOT an empty history: before the first view-model
    // arrives we have observed nothing, so the panel is left as rendered rather
    // than told there are no runs. Only a real projection writes rows.
    if (!panel || !activity) {
        return;
    }
    const list = document.getElementById('techLeadActivityList');
    if (list) {
        const focused = techLeadActivityFocusedAction(document.activeElement || null);
        if (applyTechLeadActivityRows(list, activity)) {
            restoreTechLeadActivityFocus(list, focused);
        }
    }
    const count = document.getElementById('techLeadActivityCount');
    if (count) {
        count.textContent = techLeadActivityCountText(activity);
    }
}

// Write the planned rows into ``list``. Returns whether any markup was replaced,
// which is exactly when focus may have been destroyed and needs restoring.
function applyTechLeadActivityRows(list, activity) {
    const existing = list.children ? Array.from(list.children) : [];
    const plan = techLeadActivityRowPlan(activity, existing.map(techLeadActivityNodeKey));
    if (plan.rebuild) {
        list.innerHTML = plan.html;
        return true;
    }
    let replaced = false;
    plan.replace.forEach((row, index) => {
        const node = existing[index];
        if (node && node.outerHTML !== row.html) {
            node.outerHTML = row.html;
            replaced = true;
        }
    });
    return replaced;
}

function techLeadActivityNodeKey(node) {
    return (node && node.dataset && node.dataset.tlaRow) || '';
}

// Put keyboard focus back on the equivalent action after its row was rewritten.
// A run whose row vanished loses focus to the document, which is the honest
// outcome — the control the operator was on no longer exists.
function restoreTechLeadActivityFocus(list, focused) {
    if (!focused || !list || typeof list.querySelector !== 'function') {
        return;
    }
    let slot = null;
    try {
        slot = list.querySelector(
            `[data-tla-row="${focused.row}"][data-tla-action="${focused.action}"]`,
        );
    } catch (_err) {
        // A run identity that cannot be expressed as a selector costs the focus
        // restore, never the refresh itself.
        return;
    }
    const control = slot && typeof slot.querySelector === 'function'
        ? slot.querySelector('button')
        : null;
    if (control && typeof control.focus === 'function') {
        control.focus();
    }
}

// Convenience for the load/refresh hooks.
function renderTechLeadActivityFromDashboardData() {
    updateTechLeadActivityPanel(techLeadActivityState());
}
